### PR TITLE
Docker user directives

### PR DESCRIPTION
**Pull request by mourya-33**
### Feature or Bugfix
- Bugfix

### Detail
This PR is to add USER directive to all the Docker files

### Relates
https://github.com/awslabs/aws-dataall/issues/864

### Security
Please answer the questions below briefly where applicable, or write `N/A`. Based on
[OWASP 10](https://owasp.org/Top10/en/).

- Does this PR introduce or modify any input fields or queries - this includes
fetching data from storage outside the application (e.g. a database, an S3 bucket)? No
  - Is the input sanitized? Yes
  - What precautions are you taking before deserializing the data you consume? N/A
  - Is injection prevented by parametrizing queries? N/A
  - Have you ensured no `eval` or similar functions are used? N/A
- Does this PR introduce any functionality or component that requires authorization? N/A
  - How have you ensured it respects the existing AuthN/AuthZ mechanisms? N/A
  - Are you logging failed auth attempts? N/A
- Are you using or adding any cryptographic features? N/A
  - Do you use a standard proven implementations? Yes
  - Are the used keys controlled by the customer? Where are they stored? N/A
- Are you introducing any new policies/roles/users? No
  - Have you used the least-privilege principle? How? Yes, we are disabling the containers to run as root user. This change will create a system user without root privileges to run the containers.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by dbalintx** [comment]
Left 1 comment but otherwise LGTM! How did you test it?

File: backend/docker/prod/lambda/Dockerfile [hunk 1]
```diff
@@ -14,17 +17,22 @@ RUN yum -y install openssl-devel bzip2-devel libffi-devel postgresql-devel gcc u
...
+
+USER ${CONTAINER_USER}
+
 ## Add your source
 WORKDIR ${FUNCTION_DIR}
 
-COPY backend/requirements.txt ./requirements.txt
+COPY --chown=${CONTAINER_USER}:root backend/requirements.txt ./requirements.txt
 RUN $PYTHON_VERSION -m pip install -U pip
 RUN $PYTHON_VERSION -m pip install -r requirements.txt -t .
```
Comment (dbalintx) [inline backend/docker/prod/lambda/Dockerfile:30]: when I deployed, the build for this command failed when building the image.
I think the issue is related to `FUNCTION_DIR` pointing to /home/app when it should point to the home folder of the new user you created

Comment (mourya-33) [reply backend/docker/prod/lambda/Dockerfile:30]: Resolved by adding permissions to app dir @dbalintx

Comment (mourya-33) [reply backend/docker/prod/lambda/Dockerfile:32]: Updated the permissions for FUNCTION_DIR

File: frontend/docker/prod/Dockerfile [hunk 2]
```diff
@@ -1,16 +1,42 @@
...
+
+RUN groupadd -r ${CONTAINER_USER_GROUP} && useradd -m -r -g ${CONTAINER_USER_GROUP} ${CONTAINER_USER}
+
+RUN mkdir -p /var/cache/nginx && chown -R ${CONTAINER_USER}:root /var/cache/nginx && \
+    mkdir -p /var/log/nginx  && chown -R ${CONTAINER_USER}:root /var/log/nginx && \
+    mkdir -p /var/lib/nginx  && chown -R ${CONTAINER_USER}:root /var/lib/nginx && \
+    touch /run/nginx.pid && chown -R ${CONTAINER_USER}:root /run/nginx.pid && \
+    mkdir -p /etc/nginx/templates /etc/nginx/ssl/certs && \
+    chown -R ${CONTAINER_USER}:root /etc/nginx && \
+    chmod -R 777 /etc/nginx/conf.d
```
[the remark this answers is not in the document]
Comment (mourya-33) [reply frontend/docker/prod/Dockerfile]: Removed the 777 permissions

File: backend/docker/prod/lambda/Dockerfile [hunk 3]
```diff
@@ -19,22 +22,29 @@ RUN dnf -y install \
...
+RUN chown -R ${CONTAINER_USER}:root ${FUNCTION_DIR}
+
+USER ${CONTAINER_USER}
 
 # App specific requirements
-COPY backend/requirements.txt ./requirements.txt
+COPY --chown=${CONTAINER_USER}:root backend/requirements.txt ./requirements.txt
 
 # Install App requirements
 RUN /bin/bash -c "${PYTHON_VERSION} -m pip install setuptools"
```
Comment (dlpzx) [inline backend/docker/prod/lambda/Dockerfile:39]: Why are we not adding the `--prefix /usr/local` that we add in other Dockerfiles here?

File: backend/docker/prod/lambda/Dockerfile [hunk 4]
```diff
@@ -19,22 +22,29 @@ RUN dnf -y install \
...
-COPY backend/. ./
+COPY --chown=${CONTAINER_USER}:root backend/. ./
 
 # App configuration file
 ENV config_location="config.json"
-COPY config.json ./config.json
+COPY --chown=${CONTAINER_USER}:root config.json ./config.json
 
 ## You must add the Lambda Runtime Interface Client (RIC) for your runtime.
 RUN $PYTHON_VERSION -m pip install awslambdaric --target ${FUNCTION_DIR}
```
Comment (dlpzx) [inline backend/docker/prod/lambda/Dockerfile:50]: Same why don't we add `--prefix /usr/local`? in the pip install command?

File: backend/docker/prod/ecs/Dockerfile [hunk 5]
```diff
@@ -48,26 +55,26 @@ RUN echo export PATH="\
...
-RUN /bin/bash -c "${PYTHON_VERSION} -m pip install -r cdk.requirements.txt"
+RUN /bin/bash -c "${PYTHON_VERSION} -m pip install setuptools --prefix /usr/local"
+RUN /bin/bash -c "${PYTHON_VERSION} -m pip install -r requirements.txt --prefix /usr/local"
+RUN /bin/bash -c "${PYTHON_VERSION} -m pip install -r cdk.requirements.txt --prefix /usr/local"
 
 # App code
-ADD backend/dataall /dataall
+ADD --chown=${CONTAINER_USER}:root backend/dataall /dataall
 VOLUME ["/dataall"]
 ADD backend/cdkproxymain.py /cdkproxymain.py
```
Comment (dlpzx) [inline backend/docker/prod/ecs/Dockerfile:69]: Why don't we change ownership in the line 62?

File: backend/docker/prod/lambda/Dockerfile (hunk 3 above)
Comment (mourya-33) [reply backend/docker/prod/lambda/Dockerfile:39]: This is fixed @dlpzx . Looks like this file is recently updated to have the setup tools line, the one i have on my local that i pulled a while ago did not seem to have it.

File: backend/docker/prod/ecs/Dockerfile (hunk 5 above)
Comment (mourya-33) [reply backend/docker/prod/ecs/Dockerfile]: Fixed @dlpzx

**Comment by dlpzx** [comment]
@dbalintx I have seen your deployment in AWS with the latest changes.
- [X] CICD succeeds
- [X] frontend views appear as expected
- [X] dataset creation succeeds

**Comment by dbalintx** [comment]
> @dbalintx I have seen your deployment in AWS with the latest changes. I went ahead and tested some items there:
> 
> * [x]  CICD succeeds
> * [x]  frontend views appear as expected
> * [ ]  dataset creation -->  it fails with the following error 'Error: EACCES: permission denied, mkdtemp '/tmp/jsii-kernel-XXXXXX''
> * [ ]  ML Studio user creation --> same error
> 
> image image

@mourya-33 I think you need to provide permissions to `/tmp` for the new user as these processes (and probably also further ones) are writting to it

**Comment by mourya-33** [comment]
@dbalintx I added permissions for tmp directory for all the docker files

**Comment by noah-paige** [comment]
Tested the latest changes to this PR in an AWS internet facing deployment:

- [x] CICD succeeds
- [x] frontend views appear as expected
- [ ] dataset creation --> it fails with the following error `'Error: EACCES: permission denied, mkdtemp '/tmp/jsii-kernel-XXXXXX''`
- [ ] ML Studio user creation --> same error

Getting the same errors as before with permission issues @mourya-33

**Comment by mourya-33** [comment]
@noah-paige i updated the permissions for temp dir and tested the dataset and environment creation as well.

**Comment by zsaltys** [comment]
@dlpzx is this ready to be merged or more testing is needed?

**Review by dlpzx — APPROVED** [review]
looks good! Ready to be merged

**Pull request merged**